src/hypofuzz/provider.py

```diff
@@ -26,29 +80,409 @@ def fresh_choice(
     return cast(ChoiceT, choice)
 
 
+# like choices_size, but handles ChoiceTemplate
+def _choices_size(choices: tuple[Union[ChoiceT, ChoiceTemplate], ...]) -> int:
+    return sum(
+        1 if isinstance(choice, ChoiceTemplate) else choices_size([choice])
+        for choice in choices
+    )
+
+
+class ReplayPriority(IntEnum):
+    FAILURE_SHRUNK = 0
+    FAILURE_UNSHRUNK = 1
+    COVERING = 2
+
+
+@dataclass
+class State:
+    choices: ChoicesT
+    replay_priority: Optional[ReplayPriority]
+    start_time: float
+    branches: Optional[frozenset[Branch]] = None
+    observation: Optional[Observation] = None
+    extra_queue_data: Optional[Any] = None
+
+
+# (priority, choice sequence, extra data)
+ReplayQueueElement = tuple[ReplayPriority, ChoicesT, Optional[Any]]
+
+
 class HypofuzzProvider(PrimitiveProvider):
+    add_observability_callback: ClassVar[bool] = True
+
     def __init__(
-        self, conjecturedata: Optional[ConjectureData], /, *, choices: ChoicesT
+        self,
+        conjecturedata: Optional[ConjectureData],
+        /,
+        # allow test-time override of the coverage collector.
+        collector: Optional[CoverageCollector] = None,
     ) -> None:
         super().__init__(conjecturedata)
-        self.choices = choices
-        self.index = 0
+        self.collector = collector
+        self._index = 0
+        self._replay_queue: SortedList[ReplayQueueElement] = SortedList(
+            key=lambda x: (x[0], _choices_size(x[1]))
+        )
+
+        # TODO: thread through a random instance from FuzzProcess here, or access
+        # the current BuildContext random?
+        self.random = Random()
```

Comment:
I don't think it's worth trying for determinism outside of our self-tests, and we don't currently have much of a use even there.  Let's add a `seed: int | None = None` argument if that comes up in a future PR, and delete the comment for now.